File: lib/features/achievements/models/achievement.dart
enum AchievementId { townhall18, warWarrior, mrLegend, defenseDoesntMatter }

extension AchievementIdApi on AchievementId {
  String get apiId => switch (this) {
    AchievementId.townhall18 => 'townhall_18',
    AchievementId.warWarrior => 'war_warrior',
    AchievementId.mrLegend => 'mr_legend',
    AchievementId.defenseDoesntMatter => 'defense_doesnt_matter',
  };

  static AchievementId? tryParse(String value) {
    for (final id in AchievementId.values) {
      if (id.apiId == value) return id;
    }
    return null;
  }
}

class Achievement {
  const Achievement({
    required this.id,
    required this.modelUrl,
    required this.earnedCount,
    required this.isRepeatable,
  });

  final AchievementId id;
  final String modelUrl;
  final int earnedCount;
  final bool isRepeatable;

  bool get isUnlocked => earnedCount > 0;
}

const achievementCatalogFallback = <Achievement>[
  Achievement(
    id: AchievementId.townhall18,
    modelUrl:
        'https://assets.clashk.ing/achievements/town-hall-18-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  ),
  Achievement(
    id: AchievementId.warWarrior,
    modelUrl:
        'https://assets.clashk.ing/achievements/war-champion-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  ),
  Achievement(
    id: AchievementId.mrLegend,
    modelUrl:
        'https://assets.clashk.ing/achievements/perfect-legends-day-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  ),
  Achievement(
    id: AchievementId.defenseDoesntMatter,
    modelUrl:
        'https://assets.clashk.ing/achievements/bad-legends-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  ),
];

File: expo/src/features/achievements/models/achievement.ts
export const ACHIEVEMENT_IDS = [
  'townhall_18',
  'war_warrior',
  'mr_legend',
  'defense_doesnt_matter',
] as const;

export type AchievementId = (typeof ACHIEVEMENT_IDS)[number];

export interface Achievement {
  readonly id: AchievementId;
  readonly modelUrl: string;
  readonly earnedCount: number;
  readonly isRepeatable: boolean;
}

export function isAchievementId(value: string): value is AchievementId {
  return (ACHIEVEMENT_IDS as readonly string[]).includes(value);
}

export function isAchievementUnlocked(achievement: Achievement): boolean {
  return achievement.earnedCount > 0;
}

export const ACHIEVEMENT_CATALOG_FALLBACK: readonly Achievement[] = Object.freeze([
  {
    id: 'townhall_18',
    modelUrl: 'https://assets.clashk.ing/achievements/town-hall-18-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  },
  {
    id: 'war_warrior',
    modelUrl: 'https://assets.clashk.ing/achievements/war-champion-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  },
  {
    id: 'mr_legend',
    modelUrl: 'https://assets.clashk.ing/achievements/perfect-legends-day-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  },
  {
    id: 'defense_doesnt_matter',
    modelUrl: 'https://assets.clashk.ing/achievements/bad-legends-achievement-badge.glb',
    earnedCount: 0,
    isRepeatable: true,
  },
]);
